File: extract_bodyweight_from_mdx.py
import os
import xml.etree.ElementTree as ET

def extract_bodyweight_from_mdx(directory):
    """
    Extract bodyweight from an MDX file in the specified directory
    
    Parameters:
        directory (str): The directory containing the .mdx file
    
    Returns:
        float: The extracted body weight in kilograms
    """
    # Define the path to the MDX file
    file_info = next((f for f in os.listdir(directory) if f.endswith('.mdx')), None)
    
    # Ensure there is exactly one .mdx file in the directory
    if not file_info:
        raise FileNotFoundError("No .mdx file found in the specified directory.")

    file_path = os.path.join(directory, file_info)

    # Read the XML file
    tree = ET.parse(file_path)
    root = tree.getroot()

    # Initialize the body weight variable
    body_weight = None

    # Find all "mass" elements and extract the body weight
    for mass_element in root.findall('.//mass'):
        if mass_element.get('label') == 'mTB':
            body_weight = float(mass_element.get('data')) / 1000  # Convert to kg
            break

    # Check if the body weight was found and handle the case where it was not
    if body_weight is None:
        raise ValueError("Body weight not found in the .mdx file.")

    return body_weight

File: COMAK/matlab_scripts/extract_bodyweight_from_mdx.m
function bodyWeight = extract_bodyweight_from_mdx(directory)
    % EXTRACT_BODYWEIGHT_FROM_MDX Extracts body weight from an MDX file.
    %
    % This function reads an MDX file in the specified directory and extracts the 
    % body weight value associated with the label "mTB".
    %
    % Parameters:
    %   directory (string): The directory containing the .mdx file.
    %
    % Returns:
    %   bodyWeight (double): The extracted body weight in kilograms.
    %
    % Author: Aurel Berger
    % Date: July 2024

    % Define the path to the MDX file
    file_info = dir(fullfile(directory, '*.mdx'));
    
    % Ensure there is exactly one .mdx file in the directory
    if isempty(file_info)
        error('No .mdx file found in the specified directory.');
    elseif length(file_info) > 1
        error('Multiple .mdx files found in the specified directory.');
    end

    % Read the XML file
    xmlDoc = xmlread(fullfile(file_info.folder, file_info.name));

    % Get the root element
    rootElement = xmlDoc.getDocumentElement();

    % Find all "mass" elements
    massElements = rootElement.getElementsByTagName('mass');

    % Initialize the body weight variable
    bodyWeight = [];

    % Loop through each mass element to find the total body mass
    for i = 0:massElements.getLength()-1
        massElement = massElements.item(i);
        if strcmp(char(massElement.getAttribute('label')), 'mTB')
            bodyWeight = str2double(massElement.getAttribute('data')) / 1000; % Convert to kg
            break;
        end
    end

    % Check if the body weight was found and handle the case where it was not
    if isempty(bodyWeight)
        error('Body weight not found in the .mdx file.');
    end
end
